packages/data-context/src/data/ProjectLifecycleManager.ts

```diff
@@ -272,6 +270,43 @@ export class ProjectLifecycleManager {
     })
   }
 
+  /**
+   * Sets the initial `activeBrowser` depending on these criteria, in order of preference:
+   *  1. The value of `--browser` passed via CLI.
+   *  2. The last browser selected in `open` mode (by name and channel) for this project.
+   *  3. The first browser found.
+   */
+  async setInitialActiveBrowser () {
+    if (this.ctx.coreData.cliBrowser) {
+      await this.setActiveBrowserByNameOrPath(this.ctx.coreData.cliBrowser)
+
+      // only finish if `cliBrowser` was successfully set - we must have an activeBrowser once this function resolves
+      if (this.ctx.coreData.activeBrowser) return
+    }
+
+    // lastBrowser is cached per-project.
+    const prefs = await this.ctx.project.getProjectPreferences(path.basename(this.projectRoot))
+    const browsers = await this.ctx.browser.machineBrowsers()
+
+    if (!browsers[0]) throw new Error('No browsers available in setInitialActiveBrowser, cannot set initial active browser')
+
+    this.ctx.coreData.activeBrowser = (prefs?.lastBrowser && browsers.find((b) => {
```

Comment:
Small, but all mutations of `coreData` should be done via `ctx.update`

Reply:
Noted, I was adapting the exist code. What does `ctx.update` do? The comment isn't super clear: https://github.com/cypress-io/cypress/blob/25e44ab8e6f531fe847949023f2bd9ff83ca592f/packages/data-context/src/DataContext.ts#L276-L281

Maybe we can add a lint rule to enforce this - I see lots of places where this rule is violated.

Reply:
I guess it doesn't do much right now. I believe Tim had an idea of using Immer for time travel debugging, or something along those lines. I'll check to see if an issue exists and make one if it does not. Probably shouldn't enforce a lint rule until we hook up something that provides real value.